### PR TITLE
Improve the interpolation performance and avoid unnecessary state clones

### PR DESCRIPTION
**Pull request by zhenfeizhang**
## Description

This PR 
- improve the internal API to avoid unnecessary clones of prover and verifier states during IOP
- improve the verifier complexity from O(num_vars^3) to O(num_vars^2)

with this PR 
```
Verify/ML/10            time:   [143.88 us 144.17 us 144.48 us]

Verify/ML/11            time:   [156.66 us 157.11 us 157.58 us]
```

with 3d5c5325bb4f7f669024e33c4e6666834662a153 (prior to this PR)
```
Verify/ML/10            time:   [261.08 us 261.52 us 262.06 us]
                        change: [+78.989% +80.218% +82.102%] (p = 0.00 < 0.05)

Verify/ML/11            time:   [290.51 us 291.10 us 291.71 us]
                        change: [+83.756% +85.006% +86.476%] (p = 0.00 < 0.05)
```

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by weikengchen** [comment]
I updated the code a little bit.

I would not say that the complexity reduces, but we know that the concrete efficiency generally improves. I adjusted the comments for this.

**Comment by zhenfeizhang** [comment]
> I would not say that the complexity reduces, but we know that the concrete efficiency generally improves. I adjusted the comments for this.

Is it because `get_divisor` is still a quadratic function?
We can actually use a similar method to compute divisors in linear time -- at a cost of 2 field divisions (where as the current code only has one division) per iteration. But it appears that O(n^2) primitive type multiplications is still faster than 1 field division. 

Anyway, thanks for the review, and I am okay with the changes.

**Comment by weikengchen** [comment]
Do you have the code for that? I am thinking about including a general-purpose case, even if it can be slightly slow (note that this interpolation usually is not the bottleneck).

**Comment by zhenfeizhang** [comment]
I just pushed the code for that, and I managed to avoid the extra field division I thought that I needed.
Here are the latest benchmark result 
```
Verify/ML/10            time:   [127.30 us 128.11 us 128.94 us]  
Verify/ML/11            time:   [147.69 us 148.81 us 150.05 us]              
```

**Comment by weikengchen** [comment]
Should `field_factorial` also use field elements in the middle of the computation?

**Review by tsunrise — DISMISSED** [review]
Looks good to me! I will make a new benchmark diagram later, and we can merge it then.

**Comment by weikengchen** [comment]
I will need to look at the `field_factorial` again. Please wait for me :)

**Comment by tsunrise** [comment]
> I will need to look at the `field_factorial` again. Please wait for me :)

sure, no rush

**Comment by zhenfeizhang** [comment]
> `field_factorial`

My bad. Misunderstood your point @weikengchen . fixed with https://github.com/arkworks-rs/sumcheck/pull/55/commits/e6d4b9d1c377ec9e0a20c42f069a6bc99cfc6877

File: src/ml_sumcheck/protocol/verifier.rs [hunk 1]
```diff
@@ -132,22 +132,137 @@ impl<F: Field> IPForMLSumcheck<F> {
...
+    //  denom[len-2] = (len-2) * (len-3) * ... * 2 * 1 * -1
+    // and the step before that is
+    //  denom[len-3] = (len-3) * (len-4) * ... * 2 * 1 * -1 * -2
+    //
+    // i.e., for any i, the one before this will be derived from
+    //  denom[i-1] = - denom[i] * (len-i) / i
+    //
+    // that is, we only need to store
+    // - the last denom for i = len-1, and
+    // - the ratio between current step and fhe last step, which is the
```
Comment (weikengchen) [inline src/ml_sumcheck/protocol/verifier.rs:167]:
```suggestion
    // - the ratio between the current step and the last step, which is the
```

**Pull request merged**